Shape aware functional

jamiebramwell opened the pull request:
This adds a wrapper around `Functional` to appropriately transform q-functions to reflect a shape displacement field. See documentation here: https://serac.readthedocs.io/en/feature-bramwell-shape_aware_functional/sphinx/dev_guide/functional.html#shape-aware-functional

jamiebramwell committed:
Work toward the adding a shape aware functional class

jamiebramwell committed:
Initial work on the shape-aware functional wrapper

jamiebramwell committed:
Merge branch 'develop' of github.com:LLNL/serac into feature/bramwell/shape_aware_functional

jamiebramwell committed:
The shape-aware domain integrals are now compiling

jamiebramwell committed:
Merge branch 'develop' of github.com:LLNL/serac into feature/bramwell/shape_aware_functional

jamiebramwell committed:
Merge branch 'develop' of github.com:LLNL/serac into feature/bramwell/shape_aware_functional

jamiebramwell committed:
Got the shape-aware domain and boundary integrals to compile

jamiebramwell committed:
Separate the shape FE space from the other trial spaces in the constructor for clarity

jamiebramwell committed:
Heat conduction now working with shape aware functional. For some reason, the struct version of the material integrand doesn't work yet

jamiebramwell committed:
Note the compilation error

jamiebramwell commented on src/serac/physics/heat_transfer.hpp:
```diff
@@ -426,8 +412,25 @@ class HeatTransfer<order, dim, Parameters<parameter_space...>, std::integer_sequ
+    residual_->AddDomainIntegral(Dimension<dim>{}, DependsOn<0, 1, NUM_STATE_VARS + active_parameters...>{}, integrand,
+                                 mesh_);
+
+    // This does compile
```
@samuelpmishLLNL , this is where the problem is. I can compile this using the commented out generic lambda, but not using the integrand struct. The compiler says the appropriate paren operator cannot be found.

samuelpmishLLNL replied:
Thanks for the tip, I'm pulling this branch now to take a look

samuelpmishLLNL replied:
looks like it was just missing a `const` on the paren operator (since the functor being invoked was marked `const`, it wasn't allowed to use the non-const member function).

On my mac, all tests are passing except for `boundary_cond` when run in parallel (it passes when run without mpirun).

jamiebramwell replied:
Ahhh that makes sense. Thank you for figuring this out!

samuelpmishLLNL committed:
add missing `const` on operator()

jamiebramwell committed:
Add the shape-aware QOI functional

jamiebramwell committed:
Clean up the shape helper functions and add a quadrature data container to the shape-aware functional

jamiebramwell committed:
Add conditional for passing state to the shape aware domain integrals

jamiebramwell committed:
Add shape aware functional to the solid mechanics module. All of the tests are passing except lce_Brightenti_tensile

jamiebramwell committed:
Merge branch 'develop' into feature/bramwell/shape_aware_functional

jamiebramwell committed:
Fix the quadrature data update bug

jamiebramwell commented:
I still need to clean up this PR and add some docs, but the tests all pass with `HeatTransfer` and `SolidMechanics` using the new `ShapeAwareFunctional` objects.

jamiebramwell committed:
Some clean up and removing the QOI shape-aware functional instantiation

jamiebramwell committed:
Add type checking to the QOI and non-QOI shape aware functional constructors

jamiebramwell committed:
Add some better docstrings to ShapeAwareFunctional

jamiebramwell commented:
This is all working and (mostly) clean and ready for review. I would still like to write a sphinx page on the implemented transformations, but the code itself is ready for review.

jamiebramwell committed:
Fix the doxygen warnings

jamiebramwell committed:
Add docs on shape-aware functional

jamiebramwell committed:
Merge branch 'develop' into feature/bramwell/shape_aware_functional

jamiebramwell committed:
Some more documentation and test clean up

jamiebramwell committed:
Fix the index issue

jamiebramwell committed:
make style

samuelpmishLLNL commented on src/serac/numerics/functional/shape_aware_functional.hpp:
```diff
@@ -0,0 +1,485 @@
+    static_assert(shape_space.family == Family::H1, "Only H1 spaces allowed for shape displacements");
+
+    for_constexpr<num_trial_spaces>([](auto i) {
+      static_assert(get<i>(trial_spaces).family != Family::HDIV,
+                    "Shape-aware functional not implemented for HDiv trial functions");
```
Since no part of serac currently supports Hdiv, I think there are more appropriate places to put this kind of static assertion (e.g. either `serac::Functional` and/or `serac::FiniteElementVector`)

samuelpmishLLNL commented on src/serac/numerics/functional/shape_aware_functional.hpp:
```diff
@@ -0,0 +1,485 @@
+  // du/dx' = du/dx * dx/dx'
+  //        = du/dx * (dx'/dx)^-1
+  //        = du_dx * (J)^-1
+  auto trial_derivative = dot(du_dx, inv(J));
```
it looks like this implementation will result in `inv(J)` being evaluated multiple times (once per argument).

samuelpmishLLNL commented on src/serac/physics/tests/solid_robin_condition.cpp:
```diff
@@ -73,13 +73,13 @@ void functional_solid_test_robin_condition()
+                                           auto [u, du_dxi] = displacement;
+                                           auto f           = u * 3.0 * (X[0] < 0.01);
+                                           return f;  // define a displacement-proportional traction at the support
+                                         });
```
this one might be worth a `// clang-format off/on` block!

samuelpmishLLNL reviewed (approved):
looks good, I like the latex expressions for the correction terms in the readthedocs:

<img width="818" alt="Screenshot 2024-01-19 at 8 31 06 AM" src="https://github.com/LLNL/serac/assets/61714427/0c218fd1-7fdd-4aa2-b32f-187e23abeb52">

I think there are some ways to simplify the way that the qf inputs/outputs are transformed but we can revisit that later if you want.

jamiebramwell committed:
Refactor shape-aware functional so that the jacobian calculations are only calculated once per quadrature point

jamiebramwell commented:
A few items in my last commit:

1. I moved the Jacobian calculations into a separate struct so they now only occur once per quadrature point (thanks to @samuelpmishLLNL  for pointing this out).
2. I removed the SFINAE in favor of constexpr if statements.
3. I added HCURL transformations and removed the static asserts.

white238 commented on src/docs/sphinx/dev_guide/functional.rst:
```diff
@@ -405,3 +405,70 @@ inside a ``std::function`` object with the appropriate signature. This
+address this issue, we automated these transformations in ``ShapeAwareFunctional``. This object has the same interface and integral definition routines as 
+``Functional``, but it takes the shape displacement field :math:`p` as a required trial parameter. 
+
+The integrands caputured in the ``Functional::Add****Integral`` methods do not need any further modifications to become "shape-aware" and produce the shape derivatives
```
```suggestion
The integrands captured in the ``Functional::Add****Integral`` methods do not need any further modifications to become "shape-aware" and produce the shape derivatives
```

white238 commented on src/serac/numerics/functional/tests/functional_qoi.cpp:
```diff
@@ -410,39 +411,137 @@ TEST(QoI, UsingL2)
   using parameter_space = H1<1>;
 
-  using qoi_type = serac::Functional<double(shape_space, parameter_space)>;
+  // Define the QOI type. Note that the shape aware functional has an extra template argument
```
```suggestion
  // Define the QOI type. Note that the shape-aware functional has an extra template argument
```

white238 commented on src/serac/numerics/functional/tests/functional_qoi.cpp:
```diff
@@ -410,39 +411,137 @@ TEST(QoI, UsingL2)
+  // Shape-aware functional only supports H1 and L2 trial functions
+  using parameter_space = H1<1, dim>;
+
+  // Define the QOI type. Note that the shape aware functional has an extra template argument
```
```suggestion
  // Define the QOI type. Note that the shape-aware functional has an extra template argument
```

white238 commented on src/serac/numerics/functional/shape_aware_functional.hpp:
```diff
@@ -0,0 +1,469 @@
+   *
+   * @param u The input trial function at a quadrature point
+   *
+   * @return The modified trial function adjusted for the underlying shape displacement field)
```
```suggestion
   * @return The modified trial function adjusted for the underlying shape displacement field
```

white238 commented on src/serac/numerics/functional/shape_aware_functional.hpp:
```diff
@@ -0,0 +1,469 @@
+/**
+ * @brief Compute the boundary area correction term for boundary integrals with a shape displacement field
+ *
+ * @tparam position_type The position input argument type
+ * @tparam shape_type The shape displacement input argument type
+ * @param X The input position (value and gradient)
+ * @param shape The input shape displacement (value and gradient)
+ * @return The area correction factor transforming the boundary integral from the reference domain area into the
+ * shape-displaced area
```
```suggestion
 * @tparam position_type The position input argument type
 * @tparam shape_type The shape displacement input argument type
 *
 * @param X The input position (value and gradient)
 * @param shape The input shape displacement (value and gradient)
 *
 * @return The area correction factor transforming the boundary integral from the reference domain area into the
 * shape-displaced area
```

white238 commented on src/serac/numerics/functional/shape_aware_functional.hpp:
```diff
@@ -0,0 +1,469 @@
+   * @tparam T The types of the arguments passed in
+   * @tparam wrt The index of the argument to take the gradient with respect to. Index 0 is always the shape
+   * displacement
+   * @param t The time
```
```suggestion
   *
   * @param t The time
```

white238 commented on src/serac/numerics/functional/shape_aware_functional.hpp:
```diff
@@ -0,0 +1,469 @@
+   *
+   * @note The first argument after time in the argument list is always the shape displacement field.
+   *
+   * @tparam T The types of the arguments passed in
+   * @param t The time
```
```suggestion
   * @tparam T The types of the arguments passed in
   *
   * @param t The time
```

jamiebramwell committed:
Fix the depends on bug and add a compile time check

jamiebramwell committed:
Add TODO test for when HCurl tet support is added to functional

jamiebramwell committed:
make style

jamiebramwell committed:
Update src/docs/sphinx/dev_guide/functional.rst

Co-authored-by: Chris White <white238@llnl.gov>

jamiebramwell committed:
Apply suggestions from code review

Co-authored-by: Chris White <white238@llnl.gov>

jamiebramwell committed:
Merge branch 'develop' into feature/bramwell/shape_aware_functional

jamiebramwell committed:
Fix the bad auto-style

jamiebramwell commented:
Any other thoughts on this PR? Particularly @kswartz92 and @barreracruz since you will likely be the ones using it. There is an example of what the code would look like in the docs here: https://serac.readthedocs.io/en/feature-bramwell-shape_aware_functional/sphinx/dev_guide/functional.html#shape-aware-functional

barreracruz commented:
Thank you for including the write-up. I think you don't need a parameter argument when you compute `serac_volume` and `serac_area` in the examples. Is this correct? Also, is there a reason behind defining the shape space as order 2?

kswartz92 commented:
> Any other thoughts on this PR? Particularly @kswartz92 and @barreracruz since you will likely be the ones using it. There is an example of what the code would look like in the docs here: https://serac.readthedocs.io/en/feature-bramwell-shape_aware_functional/sphinx/dev_guide/functional.html#shape-aware-functional

From a high level this looks good to me. I don't have an immediate need, but at some point it might make sense to allow different shape orders in the physics (hard-coded to linear at the moment).

jamiebramwell commented:
> Thank you for including the write-up. I think you don't need a parameter argument when you compute `serac_volume` and `serac_area` in the examples. Is this correct? Also, is there a reason behind defining the shape space as order 2?

Those examples are completely notional. I think it needs a parameter in the sense that the Functional interface requires some kind of trial function in the constructor.

There is no rationale behind a quadratic shape parameter. I just wanted something higher order to ensure that it works.
